src/display/annotation_layer.js

```diff
@@ -1832,157 +1844,235 @@ class ChoiceWidgetAnnotationElement extends WidgetAnnotationElement {
+      // consistent with other viewers such as Adobe Acrobat.
+      this.#dateTimePromise = parent.l10n.get("annotation_date_string", {
+        date: dateObject.toLocaleDateString(),
+        time: dateObject.toLocaleTimeString(),
+      });
+    }
 
-    const wrapper = document.createElement("div");
-    wrapper.classList.add("popupWrapper");
+    this.trigger = elements.flatMap(e => e.getElementsToTriggerPopup());
+    // Attach the event listeners to the trigger element.
+    for (const element of this.trigger) {
+      element.addEventListener("click", this.#boundToggle);
+      element.addEventListener("mouseenter", this.#boundShow);
+      element.addEventListener("mouseleave", this.#boundHide);
+      if (typeof PDFJSDev !== "undefined" && PDFJSDev.test("TESTING")) {
+        element.classList.add("popupTriggerArea");
+      }
+    }
 
-    // For Popup annotations we hide the entire section because it contains
-    // only the popup. However, for Text annotations without a separate Popup
-    // annotation, we cannot hide the entire container as the image would
-    // disappear too. In that special case, hiding the wrapper suffices.
-    this.hideElement = this.hideWrapper ? wrapper : this.container;
-    this.hideElement.hidden = true;
+    this.#container.hidden = true;
+    if (open) {
+      this.#toggle();
+    }
+  }
 
-    const popup = document.createElement("div");
-    popup.classList.add("popup");
+  static create(parameters) {
+    return new PopupElement(parameters);
+  }
 
-    const color = this.color;
-    if (color) {
-      // Enlighten the color.
-      const r = BACKGROUND_ENLIGHT * (255 - color[0]) + color[0];
-      const g = BACKGROUND_ENLIGHT * (255 - color[1]) + color[1];
-      const b = BACKGROUND_ENLIGHT * (255 - color[2]) + color[2];
-      popup.style.backgroundColor = Util.makeHexColor(r | 0, g | 0, b | 0);
+  render() {
+    if (this.#popup) {
+      return;
     }
 
+    const {
+      page: { view },
+      viewport: {
+        rawDims: { pageWidth, pageHeight, pageX, pageY },
+      },
+    } = this.#parent;
+    const popup = (this.#popup = document.createElement("div"));
+    popup.className = "popup";
+
+    if (this.#color) {
+      const baseColor = (popup.style.outlineColor = Util.makeHexColor(
+        ...this.#color
+      ));
+      if (typeof PDFJSDev !== "undefined" && PDFJSDev.test("MOZCENTRAL")) {
+        popup.style.backgroundColor = `color-mix(in srgb, ${baseColor} 30%, white)`;
+      } else {
+        // color-mix isn't supported in some browsers hence this version.
+        // See https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/color-mix#browser_compatibility
+        // TODO: Use color-mix when it's supported everywhere.
+        // Enlighten the color.
+        if (
+          CSS.supports("background-color", "color-mix(in srgb, red 30%, white)")
+        ) {
+          popup.style.backgroundColor = `color-mix(in srgb, ${baseColor} 30%, white)`;
+        } else {
```

Comment:
We could probably reduce a little bit of duplication here:
```suggestion
      if (
        (typeof PDFJSDev !== "undefined" && PDFJSDev.test("MOZCENTRAL")) ||
        CSS.supports("background-color", "color-mix(in srgb, red 30%, white)")
      ) {
        popup.style.backgroundColor = `color-mix(in srgb, ${baseColor} 30%, white)`;
      } else {
        // color-mix isn't supported in some browsers hence this version.
        // See https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/color-mix#browser_compatibility
        // TODO: Use color-mix when it's supported everywhere.
        // Enlighten the color.
```